devito/ir/clusters/cluster.py

```diff
@@ -51,17 +55,32 @@ def from_clusters(cls, *clusters):
         if not all(root.ispace.is_compatible(c.ispace) for c in clusters):
             raise ValueError("Cannot build a Cluster from Clusters with "
                              "incompatible IterationSpace")
+        if not all(root.properties == c.properties for c in clusters):
+            raise ValueError("Cannot build a Cluster from Clusters with "
+                             "non-homogeneous properties")
         exprs = chain(*[c.exprs for c in clusters])
         ispace = IterationSpace.union(*[c.ispace for c in clusters])
         dspace = DataSpace.union(*[c.dspace for c in clusters])
-        return Cluster(exprs, ispace, dspace)
+        return Cluster(exprs, ispace, dspace, properties=root.properties)
 
-    def rebuild(self, exprs):
+    def rebuild(self, *args, **kwargs):
         """
-        Build a new Cluster from a sequence of expressions. All other attributes
-        are inherited from ``self``.
+        Build a new Cluster from the attributes given as keywords. All other
+        attributes are taken from ``self``.
         """
-        return Cluster(exprs, self.ispace, self.dspace, self.guards)
+        # Shortcut for backwards compatibility
```

Comment:
Maybe raise a warning if this is only there for backward compatibility.

Reply:
oh no this is just an internal thing

and in fact, no worries -- this is changing again in one of the subsequent PRs (getting a proper cleanup)